lib/Finance/Quote/AlphaVantage_Batch.pm

```diff
@@ -0,0 +1,195 @@
+    my $ua = $quoter->user_agent();
+
+    my $stock_iter = natatime 100, @stocks;
+    while ( my @stocks_chunk = $stock_iter->() ) {
+
+        if ( !defined $ALPHAVANTAGE_API_KEY ) {
+            foreach my $stock (@stocks_chunk) {
+                $info{ $stock, 'success' } = 0;
+                $info{ $stock, 'errormsg' } =
+                    'Expected ALPHAVANTAGE_API_KEY to be set; get an API key at https://www.alphavantage.co';
+            }
+            next;
+        }
+
+        $url =
+              $ALPHAVANTAGE_URL
+            . '&apikey='
+            . $ALPHAVANTAGE_API_KEY
+            . '&symbols='
+            . join(',', @stocks_chunk);
+
+        my $get_content = sub {
+            sleep_before_query();
+            $reply = $ua->request( GET $url);
+
+            $code = $reply->code;
+            $desc = HTTP::Status::status_message($code);
+            $body = $reply->content;
+        };
+
+        &$get_content();
+
+        if ($code != 200) {
+            foreach my $stock (@stocks_chunk) {
+                $info{ $stock, 'success' } = 0;
+                $info{ $stock, 'errormsg' } = $desc;
+            }
+            next;
+        }
+
+        my $json_data;
+        eval {$json_data = JSON::decode_json $body};
+        if ($@) {
+            foreach my $stock (@stocks_chunk) {
+                $info{ $stock, 'success' } = 0;
+                $info{ $stock, 'errormsg' } = $@;
+            }
+            next;
+        }
+
+        my $try_cnt = 0;
+        while (($try_cnt < 5) && ($json_data->{'Information'})) {
+            # print STDERR "INFORMATION:".$json_data->{'Information'}."\n";
+            # print STDERR "ADDITIONAL SLEEPING HERE !";
+            sleep (20);
+            &$get_content();
+            eval {$json_data = JSON::decode_json $body};
+            $try_cnt += 1;
+        }
+
+        if ( !$json_data || $json_data->{'Error Message'} ) {
+            foreach my $stock (@stocks_chunk) {
+                $info{ $stock, 'success' } = 0;
+                $info{ $stock, 'errormsg' } =
+                    $json_data->{'Error Message'} || $json_data->{'Information'};
+            }
+            next;
+        }
+
+        if (!$json_data->{'Meta Data'}) {
+            foreach my $stock (@stocks_chunk) {
+                $info{ $stock, 'success' } = 0;
+                $info{ $stock, 'errormsg' } = ( $json_data->{'Information'} || "No useable data returned" ) ;
+            }
+            next;
+        }
+
+        if ( !$json_data->{'Stock Quotes'} ) {
+            foreach my $stock (@stocks_chunk) {
+                $info{ $stock, 'success' } = 0;
+                $info{ $stock, 'errormsg' } = "json_data doesn't contain Stock Quotes hash";
+            }
+            next;
+        }
+
+        foreach my $bq (@{$json_data->{'Stock Quotes'}}) {
+            foreach my $stock (@stocks_chunk) {
+                if ( $bq->{'1. symbol'} ne $stock ) {
+                    next;
+                }
```

Comment:
This has undesirably quadratic time complexity. However, my knowledge of Perl is still quite limited so this was the best that I could come up with. Edits/suggestions happily accepted 😄